task/buildah/0.4/ARTIFACT_REUSE_FEATURE.md

```diff
@@ -0,0 +1,1249 @@
+
+```mermaid
+flowchart TD
+    A[Start: Pre-build Step] --> B{REUSE_ARTIFACTS enabled?}
+    B -->|No| C[Skip reuse, proceed with fresh build]
+    B -->|Yes| D[Calculate Git Tree Hash]
+    D --> E[Check for existing tree hash tag]
+    E --> F{Tree hash tag exists?}
+    F -->|No| G[No reuse possible, proceed with fresh build]
+    F -->|Yes| H[Fetch signing key from cluster]
+    H --> I[Verify attestation signature]
+    I --> J{Attestation valid?}
+    J -->|No| K[Attestation verification failed, proceed with fresh build]
+    J -->|Yes| L[Extract tree hash from provenance]
+    L --> M{Tree hashes match?}
+    M -->|No| N[Tree hash mismatch, proceed with fresh build]
+    M -->|Yes| O[Extract build parameters from attestation]
+    O --> P[Load parameter defaults from JSON]
+    P --> Q[Compare build parameters]
+    Q --> R{All parameters match?}
+    R -->|No| S[Parameter mismatch, proceed with fresh build]
+    R -->|Yes| T[Check SBOM availability]
+    T --> U{SBOM exists on reused image?}
+    U -->|No| V[No SBOM found, proceed with fresh build]
+    U -->|Yes| W[Extract SBOM reference from attestation]
+    W --> X{SBOM reference in attestation?}
+    X -->|No| Y[No SBOM reference in attestation, proceed with fresh build]
+    X -->|Yes| Z[Download SBOM and calculate digest]
+    Z --> AA{SBOM digest matches attestation?}
+    AA -->|No| BB[SBOM integrity check failed, proceed with fresh build]
+    AA -->|Yes| CC[Reuse artifact - skip build]
+    CC --> DD[Emit REUSED_IMAGE_REF result]
+    DD --> EE[End: Artifact reused successfully]
+    
+    C --> FF[End: Fresh build]
+    G --> FF
+    K --> FF
+    N --> FF
+    S --> FF
+    V --> FF
+    BB --> FF
+    Y --> FF
+    
+    %% Color styling
+    classDef yesPath fill:#d4edda,stroke:#155724,color:#155724
+    classDef noPath fill:#f8d7da,stroke:#721c24,color:#721c24
+    classDef success fill:#d1ecf1,stroke:#0c5460,color:#0c5460
+    classDef failure fill:#f5c6cb,stroke:#721c24,color:#721c24
+    
+    %% Apply colors to yes paths (green)
+    class B,X,AA,CC,DD,EE yesPath
+    
+    %% Apply colors to no paths (red)
+    class C,G,K,N,S,V,BB,FF,Y noPath
+    
+    %% Apply colors to final outcomes
+    class EE success
+    class FF failure
+```
+
+**Decision Points Explained:**
+
+1. **REUSE_ARTIFACTS Check**: Feature can be disabled for testing/debugging
+2. **Tree Hash Calculation**: Based on actual source content, not commit metadata
+3. **Tree Hash Tag Existence**: Checks if a previous build with same content exists
+4. **Attestation Verification**: Ensures the existing artifact is authentic and signed
+5. **Tree Hash Consistency**: Verifies the attestation contains the expected tree hash
+6. **Parameter Comparison**: Compares all build parameters (using defaults when needed)
+7. **SBOM Availability**: Ensures security compliance by requiring SBOM on reused artifacts
+8. **SBOM Integrity Verification**: Verifies the attached SBOM matches the one recorded in provenance
```

Comment:
There's another one missing: compare the content of every volume mounted from the kubernetes namespace. The content of every ConfigMap, Secret etc. that gets mounted can change between runs without having an effect on any of the other decision criteria